pop/service/service.go

```diff
@@ -224,47 +466,451 @@ func (s *Service) CheckConfig(req *network.Envelope) {
 // CheckConfigReply strips the attendees missing in the reply, if the
 // PopStatus == PopStatusOK.
 func (s *Service) CheckConfigReply(req *network.Envelope) {
-	ccrVal, ok := req.Msg.(*CheckConfigReply)
-	var ccr *CheckConfigReply
-	ccr = func() *CheckConfigReply {
+	ccrVal, ok := req.Msg.(*checkConfigReply)
+	var ccr *checkConfigReply
+	ccr = func() *checkConfigReply {
 		if !ok {
 			log.Errorf("Didn't get a CheckConfigReply: %v", req.Msg)
 			return nil
 		}
-		if !bytes.Equal(ccrVal.PopHash, s.data.Final.Desc.Hash()) {
-			log.Error("Not correct hash")
+		var final *FinalStatement
+		if final, ok = s.data.Finals[string(ccrVal.PopHash)]; !ok {
+			log.Error("No party with given hash")
 			return nil
 		}
 		if ccrVal.PopStatus < PopStatusOK {
-			log.Lvl1("Wrong pop-status:", ccrVal.PopStatus)
+			log.Error("Wrong pop-status:", ccrVal.PopStatus)
 			return nil
 		}
-		s.intersectAttendees(ccrVal.Attendees)
+		final.Attendees = intersectAttendees(final.Attendees, ccrVal.Attendees)
 		return ccrVal
 	}()
-	if len(s.ccChannel) == 0 {
-		s.ccChannel <- ccr
+	if syncData, ok := s.syncs[string(ccrVal.PopHash)]; ok {
+		if len(syncData.ccChannel) == 0 {
+			syncData.ccChannel <- ccr
+		}
+	} else {
+		log.Error("No hash for sync found")
 	}
 }
 
-// Get intersection of attendees
-func (s *Service) intersectAttendees(atts []abstract.Point) {
-	na := []abstract.Point{}
-	for i, p := range s.data.Final.Attendees {
-		for _, d := range atts {
-			if p.Equal(d) {
-				na = append(na, p)
-				continue
+/* -------------Verification functions------------- */
+
+// Verification function for signing during Finalization
+func (s *Service) bftVerifyFinal(Msg []byte, Data []byte) bool {
+	final, err := NewFinalStatementFromToml(Data)
+	if err != nil {
+		log.Error(err.Error())
+		return false
+	}
+	hash, err := final.Hash()
+	if err != nil {
+		log.Error(err.Error())
+		return false
+	}
+	if !bytes.Equal(hash, Msg) {
+		log.Error("hash of received Final stmt and msg are not equal")
+		return false
+	}
+	var fs *FinalStatement
+	var ok bool
+
+	if fs, ok = s.data.Finals[string(final.Desc.Hash())]; !ok {
+		log.Error("final Statement not found")
+		return false
+	}
+
+	hash, err = fs.Hash()
+
+	if !bytes.Equal(hash, Msg) {
+		log.Error("hash of lccocal Final stmt and msg are not equal")
+		return false
+	}
+	return true
+}
+
+// Verification function for sighning during Merging
+func (s *Service) bftVerifyMerge(Msg []byte, Data []byte) bool {
+	stmtsMap, err := decodeMapFinal(Data)
+	if err != nil {
+		log.Error("VerifyMerge: can't decode Data: " + err.Error())
```

Comment:
We don't consider these messages as errors - it's a verification function, and up to some point it's normal that the verification fails - this is why we verify ;)

So change the `log.Error` to `log.Lvl2` or so. It's not an actual error, it's 'just' that the verification decided that it's bad data. The same goes for the rest of the method, replace `log.Error` with `log.Lvl2`.